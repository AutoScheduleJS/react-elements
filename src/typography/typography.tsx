import { css } from 'emotion';
import * as React from 'react';
import { merge, mergeProps, pipe } from '../util/hoc.util';
import { paletteSurface, PaletteTheme, ThemeContext } from '../util/theme';

export interface TypographyProps {
  scale: keyof TypographyScale;
  emphase?: 'high' | 'medium' | 'low';
  baselineTop?: number;
  baselineBottom?: number;
}

type TypographyPropsExtended = React.HTMLAttributes<HTMLDivElement> & TypographyProps;

export enum TypographyCase {
  Sentence,
  AllCase,
}

interface TypographyAttribte {
  typeface: string;
  weight: 300 | 400 | 500;
  size: string;
  case: TypographyCase;
  LetterSpacing: string;
}

interface TypographyScale {
  H1: TypographyAttribte;
  H2: TypographyAttribte;
  H3: TypographyAttribte;
  H4: TypographyAttribte;
  H5: TypographyAttribte;
  H6: TypographyAttribte;
  Subtitle1: TypographyAttribte;
  Subtitle2: TypographyAttribte;
  Body1: TypographyAttribte;
  Body2: TypographyAttribte;
  Button: TypographyAttribte;
  Caption: TypographyAttribte;
  Overline: TypographyAttribte;
}

export interface TypographyThemeColor {
  base: string;
  high: string;
  medium: string;
  low: string;
}

export interface TypographyTheme {
  typeface: {
    baseTypeface: string;
    titleTypeface: string;
  };
  typography: TypographyScale;
  color: TypographyThemeColor;
}

const defaultTheme = pipe(
  (theme: any) =>
    merge(
      {
        typeface: {
          baseTypeface: 'Verdana, Geneva, sans-serif',
          titleTypeface: "'Times New Roman', Times, serif",
        },
        palette: {
          surface: paletteSurface,
        },
      } as PaletteTheme & TypographyTheme,
      theme
    ),
  (theme: any): TypographyTheme => {
    const base = {
      typeface: theme.typeface.baseTypeface,
      weight: 400,
      case: TypographyCase.Sentence,
    };
    const title = { ...base, typeface: theme.typeface.titleTypeface };
    return merge(
      {
        typography: {
          H1: { ...title, weight: 300, size: '6rem', LetterSpacing: '-0.09375rem' },
          H2: { ...title, weight: 300, size: '3.75rem', LetterSpacing: '-0.03125rem' },
          H3: { ...title, size: '3rem', LetterSpacing: '0rem' },
          H4: { ...title, size: '2.125rem', LetterSpacing: '0.015625rem' },
          H5: { ...title, size: '1.5rem', LetterSpacing: '0rem' },
          H6: { ...title, weight: 500, size: '1.25rem', LetterSpacing: '0.009375rem' },
          Subtitle1: { ...base, size: '1rem', LetterSpacing: '0.009375rem' },
          Subtitle2: { ...base, weight: 500, size: '0.875rem', LetterSpacing: '0.00625rem' },
          Body1: { ...base, size: '1rem', LetterSpacing: '0.03125rem' },
          Body2: { ...base, size: '0.875rem', LetterSpacing: '0.015625rem' },
          Button: {
            ...base,
            case: TypographyCase.AllCase,
            weight: 500,
            size: '0.875rem',
            LetterSpacing: '0.046875rem',
          },
          Caption: { ...base, size: '0.75rem', LetterSpacing: '0.025rem' },
          Overline: {
            ...base,
            case: TypographyCase.AllCase,
            size: '0.625rem',
            LetterSpacing: '0.09375rem',
          },
        } as TypographyScale,
        color: {
          base: theme.palette.surface.on,
          high: theme.palette.surface.highEmphase,
          medium: theme.palette.surface.mediumEmphase,
          low: theme.palette.surface.disabled,
        },
      } as TypographyTheme,
      theme
    );
  }
);

const baselineStrut = (distance: number | undefined) => `
  display: inline-block; width: 0; content: ''; height: ${distance}px
`;

const typeScale = (
  theme: TypographyTheme,
  scale: keyof TypographyScale,
  baselineTop = 0,
  emphase: keyof TypographyThemeColor = 'high',
  baselineBottom?: number
): string => {
  const attr = theme.typography[scale];
  const bottomPos =
    baselineBottom === undefined
      ? ''
      : `
    position: absolute;
    bottom: ${baselineBottom}px;
  `;
  const color = `${theme.color.base}${theme.color[emphase]}`;
  return css`
    color: ${color};
    font-family: ${attr.typeface};
    font-weight: ${attr.weight};
    text-transform: ${attr.case === TypographyCase.Sentence ? 'initial' : 'uppercase'};
    letter-spacing: ${attr.LetterSpacing};
    font-size: ${attr.size};
    margin-bottom: ${-1 * (baselineBottom || 0)};
    &::before {
      ${baselineStrut(baselineTop)};
      vertical-align: 0;
    }
    &::after {
      ${baselineStrut(baselineBottom)};
      vertical-align: ${-1 * (baselineBottom || 0)};
    }
    ${bottomPos};
  `;
};

export const Typography: React.FunctionComponent<TypographyPropsExtended> = props => {
  const { children, scale, baselineTop, baselineBottom, emphase, ...defaultHostProps } = props;
  const theme = defaultTheme(React.useContext(ThemeContext));
  const hostProps = mergeProps(
    { className: typeScale(theme, scale, baselineTop, emphase, baselineBottom) },
    defaultHostProps
  );
  return <div {...hostProps}>{children}</div>;
};

export const TypographyProps = (options: TypographyProps & { theme?: any }) => {
  const { theme: incomingTheme, scale, baselineTop, emphase, baselineBottom } = options;
  const theme = defaultTheme(incomingTheme);
  return { className: typeScale(theme, scale, baselineTop, emphase, baselineBottom) };
};
