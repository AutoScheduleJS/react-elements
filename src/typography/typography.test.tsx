import { TypographyProps } from './typography';
import * as React from 'react';
import { render } from 'react-testing-library';

test('typography props', async () => {
  const props = TypographyProps({ scale: 'H1', emphase: 'high' });
  const { baseElement } = render(<div {...props}>test</div>);
  expect(baseElement).toMatchSnapshot('<typography props>')
});