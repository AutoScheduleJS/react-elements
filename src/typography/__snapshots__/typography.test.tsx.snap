// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography props with Body1 scale: scale: 'Body1' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0.03125rem;
  -moz-letter-spacing: 0.03125rem;
  -ms-letter-spacing: 0.03125rem;
  letter-spacing: 0.03125rem;
  font-size: 1rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Body2 scale: scale: 'Body2' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  font-size: 0.875rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Button scale: scale: 'Button' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 500;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.046875rem;
  -moz-letter-spacing: 0.046875rem;
  -ms-letter-spacing: 0.046875rem;
  letter-spacing: 0.046875rem;
  font-size: 0.875rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Caption scale: scale: 'Caption' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0.025rem;
  -moz-letter-spacing: 0.025rem;
  -ms-letter-spacing: 0.025rem;
  letter-spacing: 0.025rem;
  font-size: 0.75rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with H2 scale: scale: 'H2' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 300;
  text-transform: initial;
  -webkit-letter-spacing: -0.03125rem;
  -moz-letter-spacing: -0.03125rem;
  -ms-letter-spacing: -0.03125rem;
  letter-spacing: -0.03125rem;
  font-size: 3.75rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with H3 scale: scale: 'H3' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-size: 3rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with H4 scale: scale: 'H4' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0.015625rem;
  -moz-letter-spacing: 0.015625rem;
  -ms-letter-spacing: 0.015625rem;
  letter-spacing: 0.015625rem;
  font-size: 2.125rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with H5 scale: scale: 'H5' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-size: 1.5rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with H6 scale: scale: 'H6' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 500;
  text-transform: initial;
  -webkit-letter-spacing: 0.009375rem;
  -moz-letter-spacing: 0.009375rem;
  -ms-letter-spacing: 0.009375rem;
  letter-spacing: 0.009375rem;
  font-size: 1.25rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Overline scale: scale: 'Overline' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 400;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.09375rem;
  -moz-letter-spacing: 0.09375rem;
  -ms-letter-spacing: 0.09375rem;
  letter-spacing: 0.09375rem;
  font-size: 0.625rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Subtitle1 scale: scale: 'Subtitle1' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 400;
  text-transform: initial;
  -webkit-letter-spacing: 0.009375rem;
  -moz-letter-spacing: 0.009375rem;
  -ms-letter-spacing: 0.009375rem;
  letter-spacing: 0.009375rem;
  font-size: 1rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with Subtitle2 scale: scale: 'Subtitle2' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: Verdana,Geneva,sans-serif;
  font-weight: 500;
  text-transform: initial;
  -webkit-letter-spacing: 0.00625rem;
  -moz-letter-spacing: 0.00625rem;
  -ms-letter-spacing: 0.00625rem;
  letter-spacing: 0.00625rem;
  font-size: 0.875rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with all props: scale: 'H1', emphase: 'high', baselineTop: 10, baselineBottom: 15 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 300;
  text-transform: initial;
  -webkit-letter-spacing: -0.09375rem;
  -moz-letter-spacing: -0.09375rem;
  -ms-letter-spacing: -0.09375rem;
  letter-spacing: -0.09375rem;
  font-size: 6rem;
  margin-bottom: -15;
  position: absolute;
  bottom: 15px;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 10px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: 15px;
  vertical-align: -15;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with low emphase: scale: 'H1', emphase: 'low' 1`] = `
.emotion-0 {
  color: #FFFFFF61;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 300;
  text-transform: initial;
  -webkit-letter-spacing: -0.09375rem;
  -moz-letter-spacing: -0.09375rem;
  -ms-letter-spacing: -0.09375rem;
  letter-spacing: -0.09375rem;
  font-size: 6rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with medium emphase: scale: 'H1', emphase: 'medium' 1`] = `
.emotion-0 {
  color: #FFFFFF99;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 300;
  text-transform: initial;
  -webkit-letter-spacing: -0.09375rem;
  -moz-letter-spacing: -0.09375rem;
  -ms-letter-spacing: -0.09375rem;
  letter-spacing: -0.09375rem;
  font-size: 6rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;

exports[`Typography props with only mandatory props: scale: 'H1' 1`] = `
.emotion-0 {
  color: #FFFFFFDD;
  font-family: 'Times New Roman',Times,serif;
  font-weight: 300;
  text-transform: initial;
  -webkit-letter-spacing: -0.09375rem;
  -moz-letter-spacing: -0.09375rem;
  -ms-letter-spacing: -0.09375rem;
  letter-spacing: -0.09375rem;
  font-size: 6rem;
  margin-bottom: 0;
}

.emotion-0::before {
  display: inline-block;
  width: 0;
  content: '';
  height: 0px;
  vertical-align: 0;
}

.emotion-0::after {
  display: inline-block;
  width: 0;
  content: '';
  height: undefinedpx;
  vertical-align: 0;
}

<body>
  <div>
    <div
      class="emotion-0"
    >
      test
    </div>
  </div>
</body>
`;
